llvm/lib/SYCLLowerIR/ModuleSplitter.cpp

```diff
@@ -1112,5 +1124,68 @@ SmallVector<ModuleDesc, 2> splitByESIMD(ModuleDesc &&MD,
   return Result;
 }
 
+Error saveModuleIRInFile(Module &M, StringRef FilePath, bool OutputAssembly) {
+  int FD = -1;
+  if (std::error_code EC = sys::fs::openFileForWrite(FilePath, FD))
+    return errorCodeToError(EC);
+
+  raw_fd_ostream OS(FD, true);
+  ModulePassManager MPM;
+  ModuleAnalysisManager MAM;
+  PassBuilder PB;
+  PB.registerModuleAnalyses(MAM);
+  if (OutputAssembly)
+    MPM.addPass(PrintModulePass(OS));
+  else
+    MPM.addPass(BitcodeWriterPass(OS));
+
+  MPM.run(M, MAM);
+  return Error::success();
+}
+
+Expected<SplittedImage> saveModuleDesc(ModuleDesc &MD, std::string Prefix,
+                                       bool OutputAssembly) {
+  if (MD.isESIMD())
+    Prefix += "_esimd";
+
+  SplittedImage SI;
+
+  Prefix += OutputAssembly ? ".ll" : ".bc";
+  Error E = saveModuleIRInFile(MD.getModule(), Prefix, OutputAssembly);
+  if (E)
+    return std::move(E);
+
+  SI.ModuleFilePath = Prefix;
+  SI.Symbols = MD.makeSymbolTable();
+
+  return std::move(SI);
```

Comment:
[no `std::move` is necessary here](https://en.cppreference.com/w/cpp/language/return#Automatic_move_from_local_variables_and_parameters)

```suggestion
  return SI;
```

Reply:
Didn't know that. Thanks, done.